teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,333 @@
+---
+status: proposed
+title: Tekton Component Versioning
+creation-date: "2021-02-01"
+last-updated: "2021-02-01"
+authors:
+  - "@vinamra28"
+  - "@piyush-garg"
+  - "@SM43"
+---
+
+# TEP-0041: Tekton Component Versioning
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+  - [Creating a ConfigMap](#creating-a-configmap)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes to add versioning for all the components of Tekton in
+such a manner that any authenticated user of the cluster on which Pipelines
+is installed can get to know version of any Tekton Component. Also with
+Tekton CLI, we can get to know the version of pipelines and commands like
+install/upgrade/downgrade/reinstall a resource from catalog only if the
+newer resource version is compatible with Pipelines.
+
+## Motivation
+
+Finding the version of tekton components such as Tekton Pipelines etc which
+are currently installed on the cluster is not an easy thing specially when we
+are a non-admin user as we may not have specific permissions to view the
+version. In the current scenario we can fetch version from the controller
+deployment and for that we need:-
+
+1. Namespace in which tekton component is installed
+2. Permissions to view deployment/pod in installed namespace
+
+The above permission may not be granted to all the users and finding the version
+becomes difficult in that case. To make `Tekton Hub CLI` subcommands such as
+install, upgrade etc work efficiently it needs to know the installed Tekton
+Component version.
+
+### Goals
+
+1. All the users having access to the cluster should be able to view the version
+   of Tekton component installed using Tekton CLI.
+2. Tekton Hub CLI should be able to install/upgrade resources from catalog on the
+   basis of Tekton component installed.
+
+### Non-Goals
+
+1. How to handle versioning of components which were released prior to this
+   TEP will not be covered.
+2. Maintaining the agreed upon solution in the respective components will not
+   be covered.
+
+### Use Cases (optional)
+
+#### Tekton Hub CLI
+
+1. Tekton Hub CLI will provide an install command like `tkn hub install task buildah`.
```

Comment:
NIT: This is already available today

Reply:
yeah this is available but as of now it's not respecting the installed pipelines version. Let's suppose I have `v0.19` installed on cluster and I try to install `git-clone` task version `0.3` which requires `minPipelineVersion: v0.21` then this is possible as of now because `tkn hub cli` commands doesn't know the pipelines version and thus directly installs the task which might break the pipeline due to incompatibility.
What we are planning is that the CLI tool should throw an error whenever we try to install a resource which is not compatible with installed pipelines version.

Reply:
ack